UOP1_Project/Assets/Scripts/Characters/StateMachine/Conditions/HasReachedRoamingDestinationSO.cs

```diff
@@ -0,0 +1,27 @@
+﻿using UnityEngine;
+using UnityEngine.AI;
+using UOP1.StateMachine;
+using UOP1.StateMachine.ScriptableObjects;
+
+
+[CreateAssetMenu(fileName = "HasReachedRoamingDestination", menuName = "State Machines/Conditions/Has Reached Roaming Destination")]
+public class HasReachedRoamingDestinationSO : StateConditionSO
+{
+	protected override Condition CreateCondition() => new HasReachedRoamingDestination();
+}
+
+public class HasReachedRoamingDestination : Condition
+{
+	private NavMeshAgent _agent;
+	private float _startTime;
+
+	public override void Awake(StateMachine stateMachine)
+	{
+		_agent = stateMachine.gameObject.GetComponent<NavMeshAgent>();
+	}
+
+	protected override bool Statement()
+	{
+		return _agent == null || !_agent.hasPath;
```

Comment:
Actually, I am checking null check on agent because it is not available on all the critters (PlantCritter is static so does not need one), another option would be to add a navmeshagent to the plantcritter and given the speed will be 0 it will act as static. Do you think it would be better in term of performance than doing NPE check here and there?

Reply:
Sorry about that, but I am by passing your GetComponent() decoration by calling it from gameObject and not from StateMachine script, especially to avoid the exception throwing in case of null.

Just to understand the costly part of doing (_agent != null), I am surprised, it is costly. I guess the status of this reference is already determined (I mean we are not destroying the components at runtime, they are defined or not in the gameObject), so for me the costly operation is the GetComponent<>() 

An optimisation could be to move the null check in awake() and keep it in a local boolean since it should not change during runtime. I will push that